File: Helpers/src/main/java/com/dg/helpers/FileHelper.java
package com.dg.helpers;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 *
 * Utilities for handling Files
 */
public class FileHelper
{
    /**
     * Move a file, even across volumes.
     * Overwrites existing files.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @return true if successful
     */
    public static boolean move(File from, File to)
    {
        return move(from, to, true);
    }

    /**
     * Move a file, even across volumes.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @return true if successful
     */
    public static boolean move(File from, File to, boolean overwrite)
    {
        try
        {
            moveWithExceptions(from, to, overwrite);
            return true;
        }
        catch (IOException ignore)
        {

        }
        return false;
    }

    /**
     * Move a file, even across volumes.
     * Overwrites existing files.
     * @param from origin path
     * @param to destination path
     * @throws IOException
     */
    public static void moveWithExceptions(File from, File to) throws IOException
    {
        moveWithExceptions(from, to, true);
    }

    /**
     * Move a file, even across volumes.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @throws IOException
     */
    public static void moveWithExceptions(File from, File to, boolean overwrite) throws IOException
    {
        if (!from.exists())
        {
            throw new FileNotFoundException(String.format("'from' file was not found (%s).", from.toString()));
        }

        if (overwrite && to.exists())
        {
            if (!to.delete())
            {
                throw new IOException(String.format("'to' file was cannot be overwritten (%s).", to.toString()));
            }
        }
        if (from.renameTo(to))
        {
            return;
        }
        if (to.createNewFile())
        {
            FileChannel source = null;
            FileChannel destination = null;
            try
            {
                source = new FileInputStream(from).getChannel();
                destination = new FileOutputStream(to).getChannel();
                destination.transferFrom(source, 0, source.size());
            }
            catch (IOException ex)
            {
                //noinspection ResultOfMethodCallIgnored
                to.delete();
                throw ex;
            }
            finally
            {
                if (source != null)
                {
                    source.close();
                }
                if (destination != null)
                {
                    destination.close();
                }
                //noinspection ResultOfMethodCallIgnored
                from.delete();
            }
        }
        else
        {
            throw new IOException(String.format("'to' file was cannot be created (%s).", to.toString()));
        }
    }

    /**
     * Copy a file, even across volumes.
     * Overwrites existing files.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @return true if successful
     */
    public static boolean copy(File from, File to)
    {
        return copy(from, to, true);
    }

    /**
     * Copy a file, even across volumes.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @return true if successful
     */
    public static boolean copy(File from, File to, boolean overwrite)
    {
        try
        {
            copyWithExceptions(from, to, overwrite);
            return true;
        }
        catch (IOException ignore)
        {

        }
        return false;
    }

    /**
     * Copy a file, even across volumes.
     * Overwrites existing files.
     * @param from origin path
     * @param to destination path
     * @throws IOException
     */
    public static void copyWithExceptions(File from, File to) throws IOException
    {
        copyWithExceptions(from, to, true);
    }

    /**
     * Copy a file, even across volumes.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @throws IOException
     */
    public static void copyWithExceptions(File from, File to, boolean overwrite) throws IOException
    {
        if (!from.exists())
        {
            throw new FileNotFoundException(String.format("'from' file was not found (%s).", from.toString()));
        }

        if (overwrite && to.exists())
        {
            if (!to.delete())
            {
                throw new IOException(String.format("'to' file was cannot be overwritten (%s).", to.toString()));
            }
        }
        if (to.createNewFile())
        {
            FileChannel source = null;
            FileChannel destination = null;
            try
            {
                source = new FileInputStream(from).getChannel();
                destination = new FileOutputStream(to).getChannel();
                destination.transferFrom(source, 0, source.size());
            }
            catch (IOException ex)
            {
                //noinspection ResultOfMethodCallIgnored
                to.delete();
                throw ex;
            }
            finally
            {
                if (source != null)
                {
                    source.close();
                }
                if (destination != null)
                {
                    destination.close();
                }
            }
        }
        else
        {
            throw new IOException(String.format("'to' file was cannot be created (%s).", to.toString()));
        }
    }
}

File: Helpers/src/main/java/com/dg/helpers/FileHelper.kt
package com.dg.helpers

import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.nio.channels.FileChannel

/**
 * Utilities for handling Files
 */
object FileHelper
{
    /**
     * Move a file, even across volumes.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @return true if successful
     */
    @JvmOverloads
    fun move(from: File, to: File, overwrite: Boolean = true): Boolean
    {
        try
        {
            moveWithExceptions(from, to, overwrite)
            return true
        }
        catch (ignore: IOException)
        {

        }

        return false
    }

    /**
     * Move a file, even across volumes.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @throws IOException
     */
    @Throws(IOException::class)
    @JvmOverloads
    fun moveWithExceptions(from: File, to: File, overwrite: Boolean = true)
    {
        if (!from.exists())
        {
            throw FileNotFoundException(String.format("'from' file was not found (%s).", from.toString()))
        }

        if (overwrite && to.exists())
        {
            if (!to.delete())
            {
                throw IOException(String.format("'to' file was cannot be overwritten (%s).", to.toString()))
            }
        }
        if (from.renameTo(to))
        {
            return
        }
        if (to.createNewFile())
        {
            var source: FileChannel? = null
            var destination: FileChannel? = null
            try
            {
                source = FileInputStream(from).channel
                destination = FileOutputStream(to).channel
                destination!!.transferFrom(source, 0, source!!.size())
            }
            catch (ex: IOException)
            {

                to.delete()
                throw ex
            }
            finally
            {
                source?.close()
                destination?.close()

                from.delete()
            }
        }
        else
        {
            throw IOException(String.format("'to' file was cannot be created (%s).", to.toString()))
        }
    }

    /**
     * Copy a file, even across volumes.
     * Does not throw on failure.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @return true if successful
     */
    @JvmOverloads
    fun copy(from: File, to: File, overwrite: Boolean = true): Boolean
    {
        try
        {
            copyWithExceptions(from, to, overwrite)
            return true
        }
        catch (ignore: IOException)
        {

        }

        return false
    }

    /**
     * Copy a file, even across volumes.
     * @param from origin path
     * @param to destination path
     * @param overwrite should we overwrite existing files?
     * @throws IOException
     */
    @Throws(IOException::class)
    @JvmOverloads
    fun copyWithExceptions(from: File, to: File, overwrite: Boolean = true)
    {
        if (!from.exists())
        {
            throw FileNotFoundException(String.format("'from' file was not found (%s).", from.toString()))
        }

        if (overwrite && to.exists())
        {
            if (!to.delete())
            {
                throw IOException(String.format("'to' file was cannot be overwritten (%s).", to.toString()))
            }
        }
        if (to.createNewFile())
        {
            var source: FileChannel? = null
            var destination: FileChannel? = null
            try
            {
                source = FileInputStream(from).channel
                destination = FileOutputStream(to).channel
                destination!!.transferFrom(source, 0, source!!.size())
            }
            catch (ex: IOException)
            {

                to.delete()
                throw ex
            }
            finally
            {
                source?.close()
                destination?.close()
            }
        }
        else
        {
            throw IOException(String.format("'to' file was cannot be created (%s).", to.toString()))
        }
    }

    /**
     * Return the extension of the path, from the last '.' to end of string in the last portion of the path.
     * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string.
     * @param filePath File path to retrieve extension from
     * @return i.e. ".jpg", ".gz", ""
     */
    fun getExtension(filePath: String): String
    {
        val i = filePath.lastIndexOf('.')
        val p = Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\'))

        return if (i > p)
        {
            filePath.substring(i)
        }
        else ""

    }

    /**
     * Return the extension of the path, from the last '.' to end of string in the last portion of the path.
     * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string.
     * @param file File path to retrieve extension from
     * @return i.e. ".jpg", ".gz", ""
     */
    fun getExtension(file: File): String
    {
        return getExtension(file.name)
    }
}
